### PR TITLE
x86 pseudo for LEA is wrong

### LINKED ISSUE #16777: x86 pseudo for LEA is wrong
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Windows 10/Arch linux amd64
| File format of the file you reverse (mandatory)      | ELF
| Architecture/bits of the file (mandatory)            | x86 (32-bit)
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.5.0-git 24573 @ linux-x86-64 git.4.4.0-138-g64e6df500 commit: 64e6df5004b2718526761503a48437e82be3471e build: 2020-05-06__18:01:55

### Expected behavior

```
> e asm.pseudo = false
0x0804843c      8d45f8         lea eax, [var_8h]
```

```
> e asm.pseudo = true
0x0804843c      8d45f8         eax = var_8h
```

### Actual behavior

```
> e asm.pseudo = true
0x0804843c      8d45f8         eax = [var_8h]
```

The brackets are unneeded for pseudo, and therefore this syntax is inconsistent with arithmetic operators (which outputs like `eax = eax * edx`).

It is possible to tell the difference between a `mov` and `lea` in pseudo as the latter has no `dword` prefix, but this is still very confusing.

It was supposed to be fixed by #6618 but then reverted for #9126 which left the pseudocode part broken.

### Steps to reproduce the behavior 
Bin: crackme0x02 from [IOLI crackmes series](https://dustri.org/b/files/IOLI-crackme.tar.gz) (from #9126)

### PR DESCRIPTION
**Pull request by zawwwu**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**
PR alters pseudo code for lea operator to not include brackets. Variable resolution is preserved.

Example:
```
> e asm.pseudo = false
0x0804843c      8d45f8         lea eax, [var_8h]
```
```
> e asm.pseudo = true
0x0804843c      8d45f8         eax = var_8h
```

**Test plan**

Use e.g. crackme0x02 from IOLI crackmes series. Run ```e asm.pseudo = true``` and confirm pseudo code for lea looks as described in example.

**Closing issues**

closes #16777

### CODE REVIEW

File: libr/parse/p/parse_x86_pseudo.c [hunk 1]
```diff
@@ -485,6 +485,13 @@ static bool varsub (RParse *p, RAnalFunction *f, ut64 addr, int oplen, char *dat
 		}
 	}
 
+    // Strip brackets from lea pseudo
+	if (r_regex_match ("[a-z]{3} = \\[[a-z0-9_]+\\]$", "ne", tstr) &&
```
Comment (trufae) [inline libr/parse/p/parse_x86_pseudo.c:489]: There's no need to use a regex to achieve this, actually this was the actual behaviour.

See 17b17d91c2926e2900bfa4b34c83b88aa69bcc4b

it just a matter of reverting 06311f718b16176e8c1bbcf4ca36da23286c60a0

Comment (trufae) [reply libr/parse/p/parse_x86_pseudo.c:489]: Also, please add a test for this

Comment (zawwwu) [reply libr/parse/p/parse_x86_pseudo.c:489]: > There's no need to use a regex to achieve this, actually this was the actual behaviour.
> 
> See [17b17d9](https://github.com/radareorg/radare2/commit/17b17d91c2926e2900bfa4b34c83b88aa69bcc4b)
> 
> it just a matter of reverting [06311f7](https://github.com/radareorg/radare2/commit/06311f718b16176e8c1bbcf4ca36da23286c60a0)

Not quite sure what you mean, because simply reverting [06311f7](https://github.com/radareorg/radare2/commit/06311f718b16176e8c1bbcf4ca36da23286c60a0) indeed strips brackets, but at the same time breaks variable name resolution as mentioned in [#9126](https://github.com/radareorg/radare2/issues/9126). As I was tinkering with this issue I found out that substitution of combination `[reg +/- offset]` by variable name depends on brackets being there, so if brackets are removed **before** above mentioned combination is substituted by name of the variable, then (as far as I understood) part of varsub function that iterates over ebp/esp variables doesn't consider this a variable and thus substitution doesn't occur. If you're referring to changing disassembly completely that is
```
> e asm.pseudo = false
0x0804843c      8d45f8         lea eax, var_8h
```
```
> e asm.pseudo = true
0x0804843c      8d45f8         eax = var_8h

```
Then for sure we can try this (as I understood from comments it was tried and it worked), but in my opinion it breaks expectation of how lea source operand should look.

Comment (ret2libc) [reply libr/parse/p/parse_x86_pseudo.c:489]: I don't think we should change how `lea` looks in the disassembly, because it may be confusing, as in general it appears with `[..]`

**Comment by zawwwu** [comment]
I tinkered with that variable resolution loop I mentioned a bit more and came up with a much better solution. It works on a first glance, but I will write tests for it and update PR if it works for edge cases as well.

**Comment by XVilka** [comment]
By the way, this one would benefit from a small test.

File: test/db/anal/x86_64 [hunk 2]
```diff
@@ -3474,3 +3474,85 @@ EOF
...
+|           ; arg int64_t arg_8h @ rbp+0x8
+|           ; arg int64_t arg_bh @ rbp+0xb
+|           ; var int64_t var_4h @ rsp+0x4
+|           ; arg int64_t arg_ch @ rsp+0xc
+|           0x00000000      54             push rsp
+|           0x00000001      4889e5         mov rbp, rsp
+|           0x00000004      b402           mov ah, 2
+|           0x00000006      66b83333       mov ax, 0x3333
+|           0x0000000a      b855554444     mov eax, 0x44445555
+|           0x0000000f      48b8bebafeca.  movabs rax, 0xdeadbeefcafebabe
```
Comment (ret2libc) [inline test/db/anal/x86_64:3502]: Could you add the `movabs` op in the `ops` array in `replace` function? It should be just a matter of adding `{ "movabs", "# = #", {1, 2}},` to the array. I know it's not strictly related to your change, but at least we can create a correct test ;)

**Comment by ret2libc** [comment]
@trufae do you have changes to request? If not, I think this is good to merge.
